### ISSUE TITLE
MQTT: client_start before wifi connection established

### ISSUE THREAD

**Issue by sberryman** [nutechsoftware/AlarmDecoder-IoT#41]
Looking through the logs I see that MQTT init is done before wifi connection is established.

https://github.com/nutechsoftware/AlarmDecoder-IoT/blob/aa15c52dd445a5dc9c2e3c1ba7f4389b6717c3a6/components/ad2mqtt/ad2mqtt.cpp#L823-L824

In ser2sock I see you are creating a worker thread and checking `hal_get_network_connected()` and when watching the log output I see that starts AFTER the network connection is established.

I also noticed in your readme that the MQTT client connection is failing
https://github.com/nutechsoftware/AlarmDecoder-IoT/tree/aa15c52dd445a5dc9c2e3c1ba7f4389b6717c3a6#7-example-output-from-esp32-usb-serial-console

## log output
```shell
rst:0xc (SW_CPU_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:6076
ho 0 tail 12 room 4
load:0x40078000,len:14464
load:0x40080400,len:4000
entry 0x40080648
I (459) cpu_start: Pro cpu up.
I (459) cpu_start: Single core mode
I (467) cpu_start: Pro cpu start user code
I (467) cpu_start: cpu freq: 160000000
I (467) cpu_start: Application information:
I (468) cpu_start: Project name:     alarmdecoder_stsdk_esp32
I (473) cpu_start: App version:      1.0.6p6-149-ga2e560c
I (478) cpu_start: Compile time:     Dec  2 2021 13:15:34
I (483) cpu_start: ELF file SHA256:  04f3c9023c6d8944...
I (488) cpu_start: ESP-IDF:          4.3.0
I (492) heap_init: Initializing. RAM available for dynamic allocation:
I (498) heap_init: At 3FF80000 len 00002000 (8 KiB): RTCRAM
I (504) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (509) heap_init: At 3FFB8FB0 len 00027050 (156 KiB): DRAM
I (514) heap_init: At 3FFE0440 len 0001FBC0 (126 KiB): D/IRAM
I (520) heap_init: At 40078000 len 00008000 (32 KiB): IRAM
I (525) heap_init: At 40095248 len 0000ADB8 (43 KiB): IRAM
I (531) spi_fl!IOT: Starting AlarmDecoder AD2IoT network appliance version (AD2IOT-1093) build flag (webui)
!IOT: N (544) ESP32 with 2 CPU cores, WiFi/BT/BLE, silicon revision 1, 4MB external flash
!IOT: N (554) Initialize NVS subsystem start. Done.
!IOT: N (564) NVS usage 40.28%. Count: UsedEntries = (203), FreeEntries = (301), AllEntries = (504)
!IOT: N (564) init vpart slot 0 address 18 zones '1 18'
!IOT: N (584) init vpart slot 1 address 18 zones ''
!IOT: N (614) Mounting uSD card PASS.
!IOT: N (674) Initialize AD2 UART client using txpin(4) rxpin(36)
!IOT: N (674) Send '.' three times in the next 5 seconds to stop the init.
AD2IOT # 
!IOT: N (5674) Starting main task.
!IOT: N (5674) 'netmode' set to 'W'.
!IOT: I (5674) UARTCLI: network TCP/IP stack init start
!IOT: I (5674) UARTCLI: network TCP/IP stack init finish
!IOT: I (5684) UARTCLI: WiFi hardware init start
!IOT: I (5684) gpio: GPIO[12]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
!IOT: I (5704) wifi:wifi driver task: 3ffccf60, prio:23, stack:6656, core=0
!IOT: I (5704) system_api: Base MAC address is not set
!IOT: I (5704) system_api: read default base MAC address from EFUSE
!IOT: I (5714) wifi:wifi firmware version: c7d0450
!IOT: I (5714) wifi:wifi certification version: v7.0
!IOT: I (5724) wifi:config NVS flash: enabled
!IOT: I (5734) wifi:config nano formating: disabled
!IOT: I (5734) wifi:Init data frame dynamic rx buffer num: 32
!IOT: I (5734) wifi:Init management frame dynamic rx buffer num: 32
!IOT: I (5744) wifi:Init management short buffer num: 32
!IOT: I (5744) wifi:Init dynamic tx buffer num: 32
!IOT: I (5754) wifi:Init static rx buffer size: 1600
!IOT: I (5754) wifi:Init static rx buffer num: 10
!IOT: I (5754) wifi:Init dynamic rx buffer num: 32
!IOT: I (5764) wifi_init: rx ba win: 6
!IOT: I (5764) wifi_init: tcpip mbox: 32
!IOT: I (5774) wifi_init: udp mbox: 6
!IOT: I (5774) wifi_init: tcp mbox: 6
!IOT: I (5784) wifi_init: tcp tx win: 5744
!IOT: I (5784) wifi_init: tcp rx win: 5744
!IOT: I (5784) wifi_init: tcp mss: 1440
!IOT: I (5794) wifi_init: WiFi IRAM OP enabled
!IOT: I (5794) wifi_init: WiFi RX IRAM OP enabled
!IOT: I (5794) UARTCLI: DHCP Mode selected
!IOT: I (5804) phy_init: phy_version 4670,719f9f6,Feb 18 2021,17:07:07
!IOT: I (5894) wifi:mode : sta (98:f4:ab:25:e8:8c)
!IOT: I (5894) wifi:enable tsf
!IOT: I (5894) UARTCLI: WiFi hardware init done
!IOT: N (5904) TWILIO: Init done. Found and configured 0 virtual switches.
!IOT: N (5924) PUSHOVER: Init done. Found and configured 0 virtual switches.
!IOT: N (5924) WEBUI: service disabled.
!IOT: N (5924) MQTT init UUID: 41443245-4d42-4544-4410-98f4ab25e88c
!IOT: E (5934) TRANS_TCP: [sock=54] connect() error: Host is unreachable
!IOT: E (5934) MQTT_CLIENT: Error transport connect
!IOT: N (5974) MQTT: Init done. Found and configured 0 virtual switches.
!IOT: I (6144) wifi:new:<3,1>, old:<1,0>, ap:<255,255>, sta:<3,1>, prof:1
!IOT: I (6894) wifi:state: init -> auth (b0)
!IOT: I (6904) wifi:state: auth -> assoc (0)
!IOT: I (6914) wifi:state: assoc -> run (10)
!IOT: I (6944) wifi:connected with XXX, aid = 1, channel 3, 40U, bssid = XXX
!IOT: I (6944) wifi:security: WPA2-PSK, phy: bgn, rssi: -66
!IOT: I (6954) wifi:pm start, type: 1
!IOT: I (6954) UARTCLI: WIFI_EVENT_STA_CONNECTED
!IOT: W (6954) wifi:<ba-add>idx:0 (ifx:0, ba:fb:e4:44:6f:56), tid:0, ssn:0, winSize:64
!IOT: I (7004) wifi:AP's beacon interval = 102400 us, DTIM period = 1
!IOT: I (7894) esp_netif_handlers: sta ip: 192.168.7.XXX, mask: 255.255.255.0, gw: 192.168.7.1
!IOT: N (7894) NEW IP ADDRESS: if(sta) addr(192.168.7.XXX) mask(255.255.255.0) gw(192.168.7.1)
!IOT: N (8674) NEW IP ADDRESS: if(sta) addr(XXX) mask() gw()
AD2IOT # 
!IOT: N (10974) SER2SOCKD: Init done. Service starting.
!IOT: N (10984) AD2 State change: {"ready":true,"armed_away":false,"armed_stay":false,"backlight_on":false,"programming":false,"zone_bypassed":false,"ac_power":true}
AD2IOT #
```

**Comment by sberryman-bird** [comment]
Closing, looks like it does eventually connect.

**Comment by sberryman** [comment]
Closing, looks like it eventually connects

**Comment by f34rdotcom** [comment]
Thanks for the feedback and testing. Much appreciated!

Over the last few release I migrated away from watching a global to pause connections until the network was up. This was done to prevent crashes when I was not using the Async libraries in some of the modules or other self created race conditions. 

 After these changes now I can let MQTT manage its own connections and other places where I manage the socket I check for network before doing stuff. This network check now uses an event group bit and not a global but that is as far as I went in refactoring that code in the last release. Any code that calls hal_get_network_connected() IMHO still needs some more thought. I think some cases it is not even required any longer. It is also a good place to target if any event needs to be saved for later sending when the network returns.
